tools/congestion-model/src/strategy/traffic_light.rs

```diff
@@ -0,0 +1,228 @@
+                }
+            } else {
+                // no more receipts to execute
+                break;
+            }
+        }
+
+        let my_status = self.my_new_status(ctx);
+        ctx.current_block_info().insert(my_status);
+    }
+}
+
+impl TrafficLight {
+    fn reset_outgoing_gas_allowance(&mut self, ctx: &mut ChunkExecutionContext) {
+        self.outgoing_gas_allowance.clear();
+        for &shard in &self.all_shards {
+            let gas = match status(ctx, shard) {
+                TrafficLightStatus::Green => self.green_outgoing_receipts_gas,
+                TrafficLightStatus::Amber => self.amber_outgoing_receipts_gas,
+                TrafficLightStatus::Red { allowed_shard }
+                    if allowed_shard == self.shard_id.unwrap() =>
+                {
+                    self.red_outgoing_receipts_gas_single_shard
+                }
+                _else => 0,
+            };
+            self.outgoing_gas_allowance.insert(shard, gas);
+        }
+    }
+
+    fn forward_or_buffer(&mut self, receipt: Receipt, ctx: &mut ChunkExecutionContext) {
+        let allowance = self.outgoing_gas_allowance.get_mut(&receipt.receiver).unwrap();
+        if *allowance >= receipt.attached_gas {
+            *allowance = allowance.saturating_sub(receipt.attached_gas);
+            ctx.forward_receipt(receipt);
+        } else {
+            let q = self.outgoing_buffer[&receipt.receiver];
+            ctx.queue(q).push_back(receipt);
+        }
+    }
+
+    fn my_prev_status(&self, ctx: &mut ChunkExecutionContext) -> TrafficLightStatus {
+        status(ctx, self.shard_id.unwrap())
+    }
+
+    fn my_new_status(&self, ctx: &mut ChunkExecutionContext) -> TrafficLightStatus {
+        let incoming_gas_queued: GGas =
+            ctx.incoming_receipts().iter().map(|r| r.attached_gas).sum();
+        let incoming_bytes_queued: u64 = ctx.incoming_receipts().iter().map(|r| r.size).sum();
+        let outgoing_bytes_queued: u64 = self
+            .outgoing_buffer
+            .values()
+            .map(|&q| ctx.queue(q).iter().map(|r| r.size).sum::<u64>())
+            .sum();
+
+        // If we are reaching the memory limit, stop as much traffic as we can.
+        if incoming_bytes_queued + outgoing_bytes_queued >= self.red_queue_size_limit {
+            return self.red(ctx.block_height());
+        };
+
+        let amber = self.amber_gas_queue_limit;
+        let red = self.red_gas_queue_limit;
+        match incoming_gas_queued {
+            gas if gas < amber => TrafficLightStatus::Green,
+            gas if gas < red => TrafficLightStatus::Amber,
+            _else => self.red(ctx.block_height()),
+        }
+    }
+
+    fn red(&self, round: Round) -> TrafficLightStatus {
+        // use simple round-robin, it gives good fairness and is stateless
+        TrafficLightStatus::Red {
+            allowed_shard: self.all_shards[round as usize % self.all_shards.len()],
+        }
```

Comment:
I would consider using `round + self.shard_id % num_shards`, this way every target shard with red light will have a different allowed source shard.

Reply:
That's a really good point, thanks!